### PR TITLE
Type improvements

### PR DESCRIPTION
**Pull request by kuhnroyal**
### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dartlang.org/packages/dio)
- [x] I have searched for a similar pull request in the [project](https://github.com/flutterchina/dio/pulls) and found none
- [x] I have updated this branch with the latest `develop` to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I am adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests and they pass

This merge request fixes / refers to the following issues: ...

### Pull Request Description

General type improvements and lint fixes
Related to https://github.com/flutterchina/dio/issues/1372

### CODE REVIEW

**Comment by kuhnroyal** [comment]
@wendux @ueman @jgoyvaerts Looking for input/reviews

File: plugins/http2_adapter/lib/src/http2_adapter.dart [hunk 1]
```diff
@@ -124,15 +124,17 @@ class Http2Adapter extends HttpClientAdapter {
         }
       },
       onDone: () => sc.close(),
-      onError: (e) {
+      onError: (Object e, StackTrace stackTrace) {
```
Comment (ueman) [inline plugins/http2_adapter/lib/src/http2_adapter.dart:127]: Good catch!

File: dio/lib/src/headers.dart [hunk 2]
```diff
@@ -91,14 +91,16 @@ class Headers {
...
   /// case.
   void forEach(HeaderForEachCallback f) {
-    _map.keys.forEach((key) => f(key, this[key]!));
+    for (final key in _map.keys) {
+      f(key, this[key]!);
+    }
   }
 
   @override
   String toString() {
```
Comment (ueman) [inline dio/lib/src/headers.dart:100]: Maybe one can use `string.join` here?

File: dio/README.md [hunk 3]
```diff
@@ -311,10 +311,10 @@ The Options class describes the http request information and configuration. Each
   ValidateStatus validateStatus;
 
   /// Custom field that you can retrieve it later in [Interceptor]、[Transformer] and the   [Response] object.
-  Map<String, dynamic> extra;
+  Map<String, Object?> extra;
   
   /// Common query parameters
-  Map<String, dynamic /*String|Iterable<String>*/ > queryParameters;  
+  Map<String, Object? /*String|Iterable<String>*/ > queryParameters;
```
Comment (ueman) [inline dio/README.md:317]: Is the type hint here still needed?

**Review by ueman — APPROVED** [review]
Overall looks really good 👍

File: dio/README.md (hunk 3 above)
Comment (kuhnroyal) [reply dio/README.md:317]: I would leave it. It's just for documentation `dynamic/Object?` both give you no idea what is supported.

File: dio/lib/src/headers.dart (hunk 2 above)
Comment (kuhnroyal) [reply dio/lib/src/headers.dart:100]: Hmm yea, there is a test for this so I can try that later.

Comment (ueman) [reply dio/lib/src/headers.dart:100]: I have no strong opinion on this, I just thought that it may be easier to read.

**Comment by jgoyvaerts** [comment]
LGTM.
Probably need to do this as a 5.x release though, the dynamic to Object changes might cause issues for some users?

**Comment by kuhnroyal** [comment]
Added a `develop-5.0` branch and a project board.

**Pull request merged**